clang/lib/Driver/Driver.cpp

```diff
@@ -3904,7 +3904,10 @@ class OffloadingActionBuilder final {
       // Device compilation generates LLVM BC.
       if (CurPhase == phases::Compile) {
         for (Action *&A : SYCLDeviceActions) {
-          types::ID OutputType = types::TY_LLVM_BC;
+          types::ID OutputType = (Args.hasArg(options::OPT_S) &&
+                                  Args.hasArg(options::OPT_emit_llvm))
+                                     ? types::TY_LLVM_IR
+                                     : types::TY_LLVM_BC;
           if (SYCLDeviceOnly) {
             if (Args.hasArg(options::OPT_S))
               OutputType = types::TY_LLVM_IR;
```

Comment:
I'd suggest a rework of this nature for better readability:
```suggestion
          types::ID OutputType = types::TY_LLVM_BC;
          if ((SYCLDeviceOnly || Args.hasArg(options::OPT_emit_llvm)) &&
              Args.hasArg(options::OPT_S))
            OutputType = types::TY_LLVM_IR;
          if (SYCLDeviceOnly) {
```

As a side-note, it would be great if our options' parsing design would somehow underline that `-fsycl-device-only`, in a way, includes the `-emit-llvm` effect. Pseudo-coding a couple approaches:
1.
```
class SYCLActionBuilder {
  ...
  bool EmitLLVM = false;
  ...
}
// First code block, probably in the `initialize()` method
if (SYCLDeviceOnly || Args.hasArg(options::OPT_emit_llvm))
  EmitLLVM = true;
...
// Second code block, around the discussed segment:
types::ID OutputType = EmitLLVM ? types::TY_LLVM_IR : types::TY_LLVM_BC;
```
2.
```
if (SYCLDeviceOnly) {
  Args.push_back(Arg(options::OPT_emit_llvm)); // Input args list is likely const, though
  // Other stuff for SYCL device only
}
if (Args.hasArg(options::OPT_emit_llvm) && Args.hasArg(options::OPT_S))
  ...
```
Naturally, approach 1 looks more favorable, and before trying to refactor the code that way, one should probably investigate if we can make similar adjustments for other conditions of the SYCL flow - this could end in better code re-use and readibility. This is merely a possible direction for future - the current PR would be less cluttered as it is, or with my initial suggestion applied.